### ISSUE TITLE
Razor on linux asp.net core is very slow

### ISSUE THREAD

**Issue by rynowak** [aspnet/Mvc#6111]
_From @mytrash on June 15, 2016 19:41_

I created small asp.net core mvc project to test the speed of the asp.net application on the Linux (Ubuntu 16.04).

I created the CRUD controller that works with the "AspUser" class (stored in the PostgreSQL database). When I call the index method that shows the list of stored users the razor rendering is extremely slow. It takes 2 seconds to render the content (it is not the first call - the first call takes 8 seconds). There are 1000 users in the database. The database request is fast, but the rendering is slow.

The wrk utility wrk -c 256 -t 32 -d 10 http://localhost:5000/aspusers shows 2,6 request per second.

I made the same example in the nodejs the wrk utility shows 20 requests per second.

HW: Amd FX 8150 8 Core, 3.6 GHz, 8 GB RAM

View:

```
@model IEnumerable<PostgresTest.Models.AspUser>
@{
    ViewData["Title"] = "Index";
}

<h2>Index</h2>
<p>
    <a asp-action="Create">Create New</a>
</p>
<table class="table">
    <thead>
    <tr>
        <th>
            @Html.DisplayNameFor(model => model.Age)
        </th>
        <th>
            @Html.DisplayNameFor(model => model.FirstName)
        </th>
        <th>
            @Html.DisplayNameFor(model => model.LastName)
        </th>
        <th>
            @Html.DisplayNameFor(model => model.CreatedAt)
        </th>
        <th>
            @Html.DisplayNameFor(model => model.UpdatedAt)
        </th>
        <th></th>
    </tr>
    </thead>
    <tbody>
        @foreach (var item in Model)
        {
            <tr>
                <td>
                    @item.Age
                </td>
                <td>
                    @item.FirstName
                </td>
                <td>
                    @item.LastName
                </td>
                <td>
                    @item.CreatedAt
                </td>
                <td>
                    @item.UpdatedAt
                </td>
                <td>
                    <a asp-action="Edit" asp-route-id="@item.Id">Edit</a> |
                    <a asp-action="Details" asp-route-id="@item.Id">Details</a> |
                    <a asp-action="Delete" asp-route-id="@item.Id">Delete</a>
                </td>
            </tr>
        }
    </tbody>
</table>
```

Myth

_Copied from original issue: aspnet/Razor#790_

**Comment by rynowak** [comment]
_From @NTaylorMullen on June 15, 2016 19:46_

@SajayAntony / @mnltejaswini is this something we see in our cross plat benchmarks?

**Comment by rynowak** [comment]
_From @Eilon on June 15, 2016 21:19_

@rynowak also.

@NTaylorMullen can you start an email thread on this?

**Comment by rynowak** [comment]
_From @SajayAntony on June 16, 2016 2:9_

@mnltejaswini will be looking into this.

**Comment by rynowak** [comment]
_From @mnltejaswini on June 16, 2016 4:59_

We are trying to repro in our lab. Will get back once I get some results.

**Comment by rynowak** [comment]
_From @mnltejaswini on June 17, 2016 18:43_

I ran the scenario on Ubuntu 14.04, 4 Core , 4GB lab machine using wrk against the similar View with in memory database.

Here are the results 
Running 10s test @ http://..
  32 threads and 256 connections
Requests/sec:     21.09
Transfer/sec:     24.51MB

 We will look into the profiles to understand the behavior.

**Comment by rynowak** [comment]
_From @pranavkm on June 17, 2016 19:12_

Could you check the console logs to verify that we aren't recompiling too often?

**Comment by rynowak** [comment]
_From @mnltejaswini on June 17, 2016 19:22_

At least the profile doesn't indicate any compilation stuff. It includes processing of AnchotTagHelper like link generation.

**Comment by rynowak** [comment]
_From @mnltejaswini on June 17, 2016 19:35_

Confirmed from the console logs that the Code generation and compilation happens only once.

**Comment by rynowak** [comment]
_From @mnltejaswini on June 23, 2016 1:9_

We looked into the profiles and found that most of the cost is coming from link generation for the <a> tag helpers. We identified few optimizations to make it better and we are working on it.

**Comment by rynowak** [comment]
_From @mytrash on June 24, 2016 9:10_

Hello friends,

I played with the source code and changed the index.cshtml like this:

```
@model IEnumerable<PostgresTest.Models.AspUser>
@{
    ViewData["Title"] = "Index";
}

<h2>Index</h2>

<p>
    <a asp-action="Create">Create New</a>
</p>
<table class="table">
    <thead>
        <tr>
            <th>
                @Html.DisplayNameFor(model => model.Age)
            </th>
            <th>
                @Html.DisplayNameFor(model => model.FirstName)
            </th>
            <th>
                @Html.DisplayNameFor(model => model.LastName)
            </th>
            <th>
                @Html.DisplayNameFor(model => model.CreatedAt)
            </th>
            <th>
                @Html.DisplayNameFor(model => model.UpdatedAt)
            </th>
            <th></th>
        </tr>
    </thead>
    <tbody>
        @foreach (var item in Model)
        {
            <tr>
                <td>
                    @Html.DisplayFor(modelItem => item.Age)
                </td>
                <td>
                    @Html.DisplayFor(modelItem => item.FirstName)
                </td>
                <td>
                    @Html.DisplayFor(modelItem => item.LastName)
                </td>
                <td>
                    @Html.DisplayFor(modelItem => item.CreatedAt)
                </td>
                <td>
                    @Html.DisplayFor(modelItem => item.UpdatedAt)
                </td>
                <td>
                    <a asp-action="Edit" asp-route-id="@item.Id">Edit</a> |
                    <a asp-action="Details" asp-route-id="@item.Id">Details</a> |
                    <a asp-action="Delete" asp-route-id="@item.Id">Delete</a>
                </td>
            </tr>
        }
    </tbody>
</table>
```

The Html.DisplayFor tag helper rapidly slows the page rendering down.

**Comment by toddams** [comment]
I encounter the same issue, simple razor view without loops compile time is very low on Ubuntu 16.04

**Comment by mkArtakMSFT** [comment]
@rynowak, I've assigned this one to you, but feel free to reassign to @NTaylorMullen if you feel this is something from his area. Thanks!

**Comment by rynowak** [comment]
> I encounter the same issue, simple razor view without loops compile time is very low on Ubuntu 16.04

Hi @toddams - this is unrelated to the original issue which is about the performance of using `DisplayFor` many times in a single view. Can you please open a new issue and describe clearly what the problem is? Please include any measurements you have taken an an example of how we can reproduce the problem. Thanks!

**Comment by mkArtakMSFT** [comment]
It seems the actionable item here would be to just concentrate on the perf improvements for the `Html.DisplayFor` method. Let's consider it for the next release.

**Comment by mkArtakMSFT** [comment]
Closing this issue as the specific action is now tracked as part of a more detailed issue: https://github.com/aspnet/Mvc/issues/6928
